dlt/common/libs/pyarrow.py

```diff
@@ -654,104 +669,82 @@ def row_tuples_to_arrow(
-            logger.warning(
-                f"Field {field.name} was reflected as decimal type, but rows contains"
-                f" {py_type.__name__}. Additional cast is required which may slow down arrow table"
-                " generation."
-            )
-            float_array = pa.array(columnar_known_types[field.name], type=pa.float64())
-            columnar_known_types[field.name] = float_array.cast(field.type, safe=False)
-        if issubclass(py_type, (dict, list, set)):
+    arrow_arrays = []
+    arrow_fields = []
+    reference_row = rows[0]
+    for col_idx, (column_name, data) in enumerate(zip(columns, np.vsplit(pivoted_rows, len(pivoted_rows)))):       
+        column_schema = columns[column_name]
+        py_type = type(reference_row[col_idx])
+        if column_schema.get("data_type") == "json" and issubclass(py_type, (dict, list, set)):
             logger.warning(
-                f"Field {field.name} was reflected as JSON type and needs to be serialized back to"
+                f"Field {column_name} was reflected as JSON type and needs to be serialized back to"
                 " string to be placed in arrow table. This will slow data extraction down. You"
                 " should cast JSON field to STRING in your database system ie. by creating and"
                 " extracting an SQL VIEW that selects with cast."
             )
-            json_str_array = pa.array(
-                [
-                    (
-                        None
-                        if s is None
-                        else json.dumps(s) if not issubclass(type(s), set) else json.dumps(list(s))
-                    )
-                    for s in columnar_known_types[field.name]
-                ]
+            arrow_array = pa.array([
+                (
+                    None if s is None else 
+                    json.dumps(s) if not issubclass(type(s), set) else
+                    json.dumps(list(s))
+                )
+                for s in data.ravel()
+            ])
+
+        elif column_schema.get("data_type") in ("decimal", "wei") and issubclass(py_type, (str, float)):
+            logger.warning(
+                f"Field {column_name} was reflected as decimal type, but rows contains"
+                f" {py_type.__name__}. Additional cast is required which may slow down arrow table"
+                " generation."
             )
-            columnar_known_types[field.name] = json_str_array
+            arrow_array = pa.array(data.ravel(), type=pa.float64()).cast(get_py_arrow_datatype(column_schema, caps, tz), safe=False)
 
-    # If there are unknown type columns, first create a table to infer their types
-    if columnar_unknown_types:
-        new_schema_fields = []
-        for key in list(columnar_unknown_types):
-            arrow_col: Optional[pa.Array] = None
+        elif column_schema.get("data_type") is not None:
             try:
-                arrow_col = pa.array(columnar_unknown_types[key])
```

Comment:
you removed this case. but we must handle this. it was added because there are NULL only columns for which we are not able to infer type. and those we remove because code downstream will not handle it.

you can do `remove_null_columns` at the end. but please add a test case for that!